commands/addons/upgrade.js

```diff
@@ -0,0 +1,127 @@
+'use strict';
+
+let cli      = require('heroku-cli-util');
+let co       = require('co');
+let _        = require('lodash');
+let resolver = require('../../lib/resolve');
+let util     = require('../../lib/util');
+
+let context;
+let heroku;
+let app;
+let name;
+let plan;
+let addon;
+let service;
+
+function noPlanError () {
+  throw new Error(`Error: No plan specified.
+You need to specify a plan to move ${cli.color.yellow(name)} to.
+For example: ${cli.color.blue('heroku addons:upgrade heroku-redis:premium-0')}
+
+${cli.color.cyan('https://devcenter.heroku.com/articles/managing-add-ons')}`);
+}
+
+function handlePlanChangeAPIError (err) {
+  if (err.statusCode === 422 && err.body.message && err.body.message.startsWith('Couldn\'t find either the add-on')) {
+    return heroku.get(`/addon-services/${service}/plans`)
+    .then(plans => {
+      plans = _.sortBy(plans, 'price.cents').map(plans => plans.name);
+      throw new Error(`${err.body.message}
+
+Here are the available plans for ${cli.color.yellow(service)}:
+${plans.join('\n')}
+
+See more plan information with ${cli.color.blue('heroku addons:plans '+service)}
+
+${cli.color.cyan('https://devcenter.heroku.com/articles/managing-add-ons')}`);
+    });
+  }
+  throw err;
+}
+
+function handleAPIError (err) {
+  if (err.statusCode === 422 && err.body.id === 'multiple_matches') {
+    let example = err.body.message.split(', ')[2] || 'redis-triangular-1234';
+    throw new Error(`${err.body.message}
+
+Multiple add-ons match ${cli.color.yellow(name)}${app ? ' on '+app : ''}
+It is not clear which add-on\'s plan you are trying to change.
+
+Specify the add-on name instead of the name of the add-on service.
+For example, instead of: ${cli.color.blue('heroku addons:upgrade '+context.args.addon+' '+(context.args.plan || ''))}
+Run this: ${cli.color.blue('heroku addons:upgrade '+example + ' ' + name + ':' + plan)}
+${!app ? 'Alternatively, specify an app to filter by with '+cli.color.blue('--app') : ''}
+${cli.color.cyan('https://devcenter.heroku.com/articles/managing-add-ons')}`);
+  }
+  throw err;
+}
+
+function* run (c, h) {
+  context = c;
+  heroku  = h;
+  app     = context.app;
+  name    = context.args.addon;
+  plan    = context.args.plan;
+
+  // called with just one argument in the form of `heroku addons:upgrade heroku-redis:hobby`
```

Comment:
that's a great idea though, it could be used in a lot of places. Especially where the app can be an argument or flag (like `heroku info`, `heroku destroy`, etc)